### ISSUE TITLE
Incorrect inertia_tensor implementation

### ISSUE THREAD

**Issue by jpcornil-git** [scikit-image/scikit-image#3229]
## Description
Implementation of inertia_tensor in skimage/measure/_moments.py is not correct

Diagonal components of inertia_tensor are currently set to second order moments  (e.g. m[2,0,0] for Ixx in a 3D system) but they should be set to the sum of all 2nd order moments but the one for the corresponding axe (i.e. sum of square of the distance from corresponding axe).

```
import numpy as np
import skimage.measure as skm

obj=np.zeros((3,3,3), dtype=int)
obj[0,0,1]=obj[2,2,1]=1

props=skm.regionprops(obj)
m=props[0].moments_central

# Second order moments
m2=np.array([
   [m[2,0,0],m[1,1,0],m[1,0,1]],
   [m[1,1,0],m[0,2,0],m[0,1,1]],
   [m[1,0,1],m[0,1,1],m[0,0,2]]
])

# Compute related inertia tensor
Itensor= -m2
(Itensor[0,0], Itensor[1,1], Itensor[2,2])=(m2[1,1]+m2[2,2], m2[0,0]+m2[2,2], m2[0,0]+ m2[1,1])
```
props[0].inertia_tensor gives:
```
array([[ 1., -1., -0.],
       [-1.,  1., -0.],
       [-0., -0.,  0.]])
```
while it should be Itensor/m[0,0,0]:
```
array([[ 1., -1., -0.],
       [-1.,  1., -0.],
       [-0., -0.,  2.]])
```
Given m2:
```
array([[ 2.,  2.,  0.],
       [ 2.,  2.,  0.],
       [ 0.,  0.,  0.]])
```

**Comment by jni** [comment]
@jpcornil-git oops, I must have made a mistake when converting to nD in #2603! (Are you using skimage 0.14?) Do you have a good reference for me to check the definition in 3D? I wasn't able to find one at the time...

**Comment by jpcornil-git** [comment]
@jni Yes, I'm using 0.14.

You should probably only replace:
`d[:] = mu[corners2] / mu0`
with:
`d[:] = (mu[corners2].sum() - mu[corners2])/mu0`

This reference may be helpful: https://ocw.mit.edu/courses/aeronautics-and-astronautics/16-07-dynamics-fall-2009/lecture-notes/MIT16_07F09_Lec26.pdf

Cheers
jpc

**Comment by jni** [comment]
Haha I see what happened now. I was looking at the previous code, which was 2D only *and* used `xy` coordinates. So `I_00` was equal to `mu[2, 0]`... This just happened to be exactly what you suggest — in `xy` coordinates — but also happened to *look* exactly like my mistaken interpretation/generalisation, in `rc` coordinates.

Working on fixing this now. Thanks so much for the report and for your patience!

**Comment by jpcornil-git** [comment]
Thanks for the update jni (not sure I understand why you are differentiating `xy` and `rc` coordinates here as these shouldn't affect diagonal components given they are all **2)

Cheers
jpc

**Comment by jpcornil-git** [comment]
Ah, I see your point, rc is not cr ... just like endianess, the one who invented the second system should be fired :-) [integrated drawbacks offsetting benefits by far] !
Cheers,
jpc

**Comment by jpcornil-git** [comment]
Hi jni,

Not sure I understood correctly your question but eigenvectors will give you the main object axes and eigenvalues inertia around each axe => eigenvector corresponding to the smallest eigenvalue will indicate the axe along the 'longest side' of the object.

Cheers
jpc

**Comment by jni** [comment]
@jpcornil-git got it, thanks! Your explanation clarified quite a few things for me.

Yes, I'm trying to determine the orientation of the major (longest) axis of the object. To get deeper at my question, I'm trying to understand this implementation:

https://github.com/scikit-image/scikit-image/blob/dc8b56f8df6b9662f072c4cb5548939f12070c7c/skimage/measure/_regionprops.py#L273-L282

because things are getting transposed in weird ways when I fix the inertia tensor, and I want to make sure I don't break another thing while fixing this. Specifically, I don't understand the factor of -0.5. I presume it has something to do with the fact that the values in the matrix are sums of squares. I could just use np.arctan2 on the values of the eigenvector, but I presume that would be much slower, so if you can explain the formula above for me, I would be most grateful!

**Comment by jpcornil-git** [comment]
Hi jni,

The -0.5 is just the consequence of axis rotation, e.g. if you rotate axis by α the 2D Inertia matrix becomes:
Ix'x' = (Ixx +Iyy)/2 + (Ixx −Iyy)/2 * cos2α − Ixy * sin2α
Iy'y' = (Ixx +Iyy)/2 + (Ixx −Iyy)/2 * cos2α + Ixy * sin2α
Ix'y' = (Ixx −Iyy)/2 * sin2α +Ixy * cos2α

=> Principal axis is when Ix'y'/last line above becomes 0 or
tan2α=-2 * Ixy/(Ixx −Iyy)

This is the same as the orientation of one of the eigenvector

```
>>> import numpy as np
>>> import numpy.linalg as npla
>>> I=np.array([[1,2],[2,3]])
>>> IEval, IEvec=npla.eig(I)
>>> IEval
array([-0.23606798,  4.23606798])
>>> IEvec
array([[-0.85065081, -0.52573111],
       [ 0.52573111, -0.85065081]])
>>> math.atan2(IEvec[1][1],IEvec[1][0])
-1.0172219678978514
>>> a,b,c=I[0,0],I[0,1],I[1,1]
>>> -0.5 * math.atan2(-2 * b, (a - c))
1.0172219678978514
```

Now, the atan formula has two solutions for α and the `orientation ` function only picks one of the two, i.e. not the one related to the major axis. You can improve this with the eigen values/vectors.

The negative sign in front of the atan2 is strange .... likely another consequence of xy/rc/... permutation/direction :-)

Cheers
jpc

**Comment by jpcornil-git** [comment]
... minus sign may also just be a consequence/fix of the Ixx<->Iyy (a<->c inversion)
Cheers
jpc

**Comment by jni** [comment]
@jpcornil-git Neat! Thanks so much, that was crystal clear!

Did you pick the right eigenvector though? You want the eigenvector corresponding to the smallest eigenvalue, which is column 0 of `I`...? Therefore:

```python
In [1]: import numpy as np
In [2]: I = np.array([[1, 2], [2, 3]])
In [3]: Λ, V = np.linalg.eig(I)
In [4]: Λ
Out[4]: array([-0.23606798,  4.23606798])

In [5]: V
Out[5]: 
array([[-0.85065081, -0.52573111],
       [ 0.52573111, -0.85065081]])

In [7]: np.arctan2(I[1, 0], I[0, 0])
Out[7]: 1.1071487177940904

In [8]: a, b, b, c = I.flat
In [9]: -0.5 * np.arctan2(-2 * b, a - c)
Out[9]: 1.0172219678978514
```

Anyway, armed with this new understanding I think I can finally untangle this. Hold, please!

**Comment by jni** [comment]
*sigh* it's worse than I thought, but it finally makes sense. If you use "xy", which is not actually the Cartesian xy with the origin at the bottom left, but rather a hybrid abomination where "x" is "columns right from the top left corner" and "y" is "rows down from the top left corner", then the "natural" direction for arctan is clockwise. But the documentation specifies counter-clockwise orientation to match common understanding of rotation, *even though that makes no sense in this coordinate system.* *That's* why there's a minus in front.

Of course the consequence of this is that you need a minus for the old behaviour but not a minus for the new one. :man_facepalming:

**Comment by jpcornil-git** [comment]
@jni 

I just picked the eigenvector matching the arctan2 formula (note that line 7 in your ipython code should use V not I, i.e. it should be `np.arctan2(V[1, 1], V[1, 0])`)

Cheers
jpc

**Comment by jni** [comment]
@jpcornil-git Ah, yes, but it should be `V[0, 0], V[1, 0]`, no? The 0th column to correspond to the 0th eigenvalue?

**Comment by jpcornil-git** [comment]
V[0] and V[1] are the two eigenvectors (and orthogonal to each other's)

```
In [52]: V[0]
Out[52]: array([-0.85065081, -0.52573111])
In [53]: V[1]
Out[53]: array([ 0.52573111, -0.85065081])
In [54]: np.dot(V[0],V[1])
Out[54]: 0.0
```
Cheers
jpc

**Comment by jpcornil-git** [comment]
... and Λ are the eigenvalues

Cheers
jpc

**Comment by jni** [comment]
@jpcornil-git from the NumPy docs:

```
v : (..., M, M) array
    The normalized (unit "length") eigenvectors, such that the
    column ``v[:,i]`` is the eigenvector corresponding to the
    eigenvalue ``w[i]``.
```

**Comment by jni** [comment]
Ok fix submitted in #3303. Thank you so much for the report and for all your help, @jpcornil-git!

**Comment by jpcornil-git** [comment]
Yep, you are right :-) !
Cheers
jpc
